File: server/src/models/AgentJob.js
import mongoose from 'mongoose';

const agentJobSchema = new mongoose.Schema({
  taskId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Task',
    required: true
  },
  agentType: {
    type: String,
    enum: ['data_extraction', 'financial_analysis', 'news_summarization', 'analyst_support', 'recommender'],
    required: true
  },
  status: {
    type: String,
    enum: ['queued', 'running', 'completed', 'failed', 'cancelled'],
    default: 'queued'
  },
  input: {
    type: Map,
    of: mongoose.Schema.Types.Mixed,
    required: true
  },
  output: {
    type: Map,
    of: mongoose.Schema.Types.Mixed,
    default: new Map()
  },
  parameters: {
    type: Map,
    of: mongoose.Schema.Types.Mixed,
    default: new Map()
  },
  priority: {
    type: Number,
    default: 0 // Higher number = higher priority
  },
  retryCount: {
    type: Number,
    default: 0
  },
  maxRetries: {
    type: Number,
    default: 3
  },
  startedAt: {
    type: Date
  },
  completedAt: {
    type: Date
  },
  estimatedDuration: {
    type: Number, // in milliseconds
    default: 60000 // 1 minute default
  },
  actualDuration: {
    type: Number // in milliseconds
  },
  error: {
    message: String,
    stack: String,
    timestamp: Date,
    retryable: { type: Boolean, default: true }
  },
  metadata: {
    type: Map,
    of: mongoose.Schema.Types.Mixed,
    default: new Map()
  },
  logs: [{
    level: {
      type: String,
      enum: ['debug', 'info', 'warn', 'error'],
      default: 'info'
    },
    message: String,
    timestamp: { type: Date, default: Date.now },
    data: Map
  }]
}, {
  timestamps: true
});

// Indexes for performance
agentJobSchema.index({ taskId: 1, agentType: 1 });
agentJobSchema.index({ status: 1, priority: -1, createdAt: 1 });
agentJobSchema.index({ createdAt: -1 });

// Virtual for duration calculation
agentJobSchema.virtual('duration').get(function() {
  if (this.startedAt && this.completedAt) {
    return this.completedAt - this.startedAt;
  }
  return null;
});

// Method to start job
agentJobSchema.methods.start = function() {
  this.status = 'running';
  this.startedAt = new Date();
  return this.save();
};

// Method to complete job
agentJobSchema.methods.complete = function(output, metadata = {}) {
  this.status = 'completed';
  this.completedAt = new Date();
  this.actualDuration = this.duration;
  this.output = new Map(Object.entries(output));
  this.metadata = new Map(Object.entries(metadata));
  return this.save();
};

// Method to fail job
agentJobSchema.methods.fail = function(error, retryable = true) {
  this.status = 'failed';
  this.completedAt = new Date();
  this.actualDuration = this.duration;
  this.error = {
    message: error.message,
    stack: error.stack,
    timestamp: new Date(),
    retryable
  };
  return this.save();
};

// Method to add log entry
agentJobSchema.methods.addLog = function(level, message, data = {}) {
  this.logs.push({
    level,
    message,
    timestamp: new Date(),
    data: new Map(Object.entries(data))
  });
  return this.save();
};

// Method to retry job
agentJobSchema.methods.retry = function() {
  if (this.retryCount < this.maxRetries && this.error?.retryable) {
    this.retryCount += 1;
    this.status = 'queued';
    this.error = undefined;
    this.startedAt = undefined;
    this.completedAt = undefined;
    return this.save();
  }
  return Promise.reject(new Error('Job cannot be retried'));
};

// Method to update status
agentJobSchema.methods.updateStatus = function(status) {
  this.status = status;
  if (status === 'running' && !this.startedAt) {
    this.startedAt = new Date();
  } else if (status === 'completed' || status === 'failed') {
    this.completedAt = new Date();
  }
  return this.save();
};

export default mongoose.model('AgentJob', agentJobSchema);

File: server/src/models/AgentJob.py
from mongoengine import Document, StringField, ReferenceField, DictField, ListField, DateTimeField, IntField, FloatField, BooleanField, EmbeddedDocument, EmbeddedDocumentField
import datetime

class LogEntry(EmbeddedDocument):
    level = StringField(choices=['debug', 'info', 'warn', 'error'], default='info')
    message = StringField()
    timestamp = DateTimeField(default=datetime.datetime.utcnow)
    data = DictField(default={})

class ErrorInfo(EmbeddedDocument):
    message = StringField()
    stack = StringField()
    timestamp = DateTimeField()
    retryable = BooleanField(default=True)

class AgentJob(Document):
    taskId = StringField(required=True)  # ReferenceField('Task') can be used if Task is defined
    agentType = StringField(required=True, choices=['data_extraction', 'financial_analysis', 'news_summarization', 'analyst_support', 'recommender'])
    status = StringField(choices=['queued', 'running', 'completed', 'failed', 'cancelled'], default='queued')
    input = DictField(required=True)
    output = DictField(default={})
    parameters = DictField(default={})
    priority = IntField(default=0)
    retryCount = IntField(default=0)
    maxRetries = IntField(default=3)
    startedAt = DateTimeField()
    completedAt = DateTimeField()
    estimatedDuration = IntField(default=60000)  # ms
    actualDuration = IntField()
    error = EmbeddedDocumentField(ErrorInfo)
    metadata = DictField(default={})
    logs = ListField(EmbeddedDocumentField(LogEntry))
    createdAt = DateTimeField(default=datetime.datetime.utcnow)
    updatedAt = DateTimeField(default=datetime.datetime.utcnow)

    meta = {
        'indexes': [
            {'fields': ['taskId', 'agentType']},
            {'fields': ['status', '-priority', 'createdAt']},
            {'fields': ['-createdAt']}
        ]
    }

    @property
    def duration(self):
        if self.startedAt and self.completedAt:
            return (self.completedAt - self.startedAt).total_seconds() * 1000
        return None

    def start(self):
        self.status = 'running'
        self.startedAt = datetime.datetime.utcnow()
        self.save()

    def complete(self, output, metadata=None):
        self.status = 'completed'
        self.completedAt = datetime.datetime.utcnow()
        self.actualDuration = self.duration
        self.output = output
        self.metadata = metadata or {}
        self.save()

    def fail(self, error, retryable=True):
        self.status = 'failed'
        self.completedAt = datetime.datetime.utcnow()
        self.actualDuration = self.duration
        self.error = ErrorInfo(
            message=getattr(error, 'message', str(error)),
            stack=getattr(error, 'stack', ''),
            timestamp=datetime.datetime.utcnow(),
            retryable=retryable
        )
        self.save()

    def add_log(self, level, message, data=None):
        entry = LogEntry(level=level, message=message, timestamp=datetime.datetime.utcnow(), data=data or {})
        self.logs.append(entry)
        self.save()

    def retry(self):
        if self.retryCount < self.maxRetries and (self.error is None or self.error.retryable):
            self.retryCount += 1
            self.status = 'queued'
            self.error = None
            self.startedAt = None
            self.completedAt = None
            self.save()
        else:
            raise Exception('Job cannot be retried')

    def update_status(self, status):
        self.status = status
        if status == 'running' and not self.startedAt:
            self.startedAt = datetime.datetime.utcnow()
        elif status in ['completed', 'failed']:
            self.completedAt = datetime.datetime.utcnow()
        self.save()
